core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -869,7 +876,19 @@ private[spark] class BlockManager(
           blockInfoManager.unlock(blockId)
         }
       } else {
-        blockInfoManager.removeBlock(blockId)
+        // If an exception was thrown then it's possible that the code in `putBody` has already
+        // notified the master about the availability of this block, so we need to send an update
+        // to remove this block location.
+        removeBlockInternal(
+          blockId, tellMaster = tellMaster && putBlockInfo.tellMaster && exceptionWasThrown)
+        if (exceptionWasThrown) {
+          // The `putBody` code may have also added a new block status to TaskMetrics, so we need
+          // to cancel that out by overwriting it with an empty block status. We only do this if
+          // the finally block was entered via an exception because doing this unconditionally would
+          // cause us to send empty block statuses for every block that failed to be cached due to
+          // a memory shortage (which is an expected failure, unlike an uncaught exception).
+          addUpdatedBlockStatusToTaskMetrics(blockId, BlockStatus.empty)
+        }
```

Comment:
nit: might be nice to put all the code relevant to exceptionWasThrown in this if